### ISSUE TITLE
Bloom

### ISSUE THREAD

**Issue by Paril** [skullernet/q2pro#362]
Hey skullernet; I did an implementation of bloom in our fork, I just pushed up an initial version of it to our fork ( https://github.com/Paril/q2pro/commit/29abf844cbf884993fe4409d6176a3bde8703ad5 ). Curious if you're interested in it and potentially helping with a few of the remaining issues on the code since you know your way around GL way more than I do. If not no worries, it's a bit of a niche feature and glowmaps look pretty good without it anyways.

It's an extension of glowmaps; the bloom only applies to them, working just like re-release (so it's effectively artist-controllable bloom) and helps soften out the glows.

![quake001](https://github.com/user-attachments/assets/2b970604-a8f5-44c7-9d6d-5af5746ca109)
![quake002](https://github.com/user-attachments/assets/e73688e6-61de-4450-a509-3b9ec9f940d1)
![quake003](https://github.com/user-attachments/assets/cac44aee-7e1d-4c6e-9457-a3708147ae03)
![quake004](https://github.com/user-attachments/assets/47d7a27e-2a26-411a-be63-345103157a85)

So far here's what I have:
* gl_bloom cvar for on/off
* gl_bloom_scale cvar for controlling the downsample size (default 0.25)
* gl_bloom_radius cvar for changing the blur radius (default 2)
* uses a second fragment output to write the bloom to a secondary texture; the bloom texture is then downsampled, and uses two-pass gaussian blurring
* works with water warp
and here's what I don't have:
* code is messy; probably can do with some re-usable functions for framebuffers especially
* does not work with sizedown/sizeup yet, but that should be trivial
* certain objects like particles write black to the bloom buffer when they should be ignored; I think we just need another shader bit to tell it to ignore the bloom output entirely on certain elements
* blur gaussian routine is hardcoded to 13 tap, but I have 5 and 9 tap available
* multiple iterations; right now the blur is only done once, so you're limited to a small radius otherwise you get some nasty artifacting

**Comment by skullernet** [comment]
Thanks. I've pushed some WIP code based on yours to `bloom` branch. It mostly works, but I'm not quite satisfied with it yet so it might be rewritten again. In my code the only controllable parameter is `gl_bloom`, which sets the number of iterations. Value 3 seems to produce results the most close to re-release.

Compared to your version:
* fixes artifacts on particles
* twice as fast
* should work with GL ES 3
* should work with sizedown/sizeup

**Comment by Paril** [comment]
Oh cool. I've pushed some new stuff recently as well, which also fixes the particle artifacting and adds bloom to color shells & beams to match re-release. Not sure if you got to those two yet. I'm running out of shader bits though so that's going to be a problem down the road lol

**Comment by skullernet** [comment]
> The other main change is I changed the way the blurring is calculated to be closer to Q2RTX, and the output quality is much higher with less artifacting.

Looks good! Although I only get around 50 FPS at the start of base2 with your latest code (with bloom disabled I get 600). Maybe default parameters of blur filter need to be tweaked?

I've imported dynamic blur filter parameter generation into my code, but I let the user specify blur radius/sigma manually (it's not resolution dependent). Also added bloom for shells/lasers, fog attenuation, etc.

> I took a different approach with separable blend funcs and always use additive on the blend output

Yeah, but those are for GL ≥ 4.0. I'd still like to support bloom with GL 3.1, so probably won't be using those.

**Comment by res2k** [comment]
> but I let the user specify blur radius/sigma manually (it's not resolution dependent).

Two notes here:
* Having the blur radius be based on the screen resolution is, in a way, more resolution independent - with a relative size, if you change your display resolution, the bloom effect size stays relatively the same. With an absolute size, it appears smaller the higher your resolution gets. (If you pick any defaults that "look like rerelease", that would only apply if using the exact same screen resolution as you do.)
* In practice, sigma and radius are actually linked - see https://en.wikipedia.org/wiki/Normal_distribution.
Changing them independently may result in scenarios where you'll have a bunch of samples with negligible impact on the result (radius too large for sigma) or you're having not enough samples (radius too small).

**Comment by Paril** [comment]
Fair enough with GL >= 4.0 - but check that the bloom doesn't break with muzzleflashes, that's where I noticed the biggest issue with the current approach (muzzleflashes have a large blank alpha area, so without being additive they just overwrote the bloom behind it). Technically separable blend funcs have been available since 2.0 as an extension, I think it's safe to assume every card has it at this point.

**Comment by Paril** [comment]
Cool, I merged that all in. Looks good!
The only thing I notice is that the 2D beams are missing glow; not sure if that's intentional or just because visually they already are kinda bloomy.

**Comment by res2k** [comment]
> I'm more concerned about good performance on wide range of GPUs rather than consistent look at different resolutions.

I'm not sure there's necessarily a contradiction here.
There's a common understanding that "higher resolution means less performance". Scaling the radius fits in there as well.
So if bloom performance becomes an issue, but it's something that's desired, lowering the resolution would improve performance - as typically expected.
I believe there's also the common assumption that, when you change the resolution in a 3D game, everything _scales_ proportionally. Making the radius dependent on resolution would thus probably fulfill user expectations.

> Maybe size of intermediate scaling buffer could be fixed instead. Will need to experiment more with it.

For any fixed size, there's probably _one_ screen resolution it works well for, but for higher resolutions the fixed size is probably not enough, and for lower resolution it's probably too much.

> > In practice, sigma and radius are actually linked
> 
> Yeah, I figured that out. But let's assume these parameters are for someone who knows what they are doing. Perhaps it needs to be more explicitly spelled in documentation.

If you spell out an explicit relation in the documentation, it seems more convenient to just automatically apply that recommendation via code. Or, if you want to make the radius configurable, make the _factor_ in `radius = factor * sigma` configurable. (I consider myself someone who knows what I'm doing, yet I also like convenience and not having to mess with multiple parameters.)

**Comment by skullernet** [comment]
> The only thing I notice is that the 2D beams are missing glow; not sure if that's intentional or just because visually they already are kinda bloomy.

I've added bloom to 2D beams as well.

> res2k also mentioned in a discussion we had that it's possible that the bloom code is supposed to be doing a box filter when doing the downsampling, so that you don't lose any bloom information when doing the downsample. 

So should we add box blur to downsampling stage? Or replace gaussian blur with box blur entirely?

> Making the radius dependent on resolution would thus probably fulfill user expectations.

In principle yes (but definitely not using @Paril's original formula as it made performance abysmal for me). What I don't like about this approach is that it requires shader recompilation each time resolution changes. This can be very annoying when e.g. resizing the window.

>  but for higher resolutions the fixed size is probably not enough, and for lower resolution it's probably too much.

Can't say I really agree. 960x540 fixed size buffer is enough for 4K. For lower resolutions it doesn't really become counter-productive unless you run the game at less than 960x540, which is highly unlikely.

> If you spell out an explicit relation in the documentation, it seems more convenient to just automatically apply that recommendation via code. 

I've removed radius parameter and changed radius to always be 2σ in current version.

**Comment by res2k** [comment]
> So should we add box blur to downsampling stage? Or replace gaussian blur with box blur entirely?

The idea was to add that to the downsampling stage.

> Can't say I really agree. 960x540 fixed size buffer is enough for 4K. For lower resolutions it doesn't really become counter-productive unless you run the game at less than 960x540, which is highly unlikely.

OTOH, quartering for _all_ resolutions would mean you'd gain some performance on lower resolutions...

**Comment by Paril** [comment]
I assume the quartering is what makes the box blur "lossless" too, since every pixel just becomes 4 pixels from the input. Not certain though.

Also I discovered one other minor bug while fixing some stuff in q2pro-rerelease: the Player Setup menu renders a bit funny now.

**Comment by skullernet** [comment]
> OTOH, quartering for all resolutions would mean you'd gain some performance on lower resolutions...

OK, I've restored resolution quartering and added dynamic bloom buffer resizing. Maybe shader recompilation on resolution change is not that bad. I've tried passing gaussian blur filter parameters via uniform buffer to avoid recompilation but that turned out to be 30% slower than having coefficients inline in shader.

> The idea was to add that to the downsampling stage.

Not sure if I implemented it correctly but indeed with box filter it shows less bloom artifacts now when rendering at very low resolutions such as 640x480. On higher resolutions I don't see any difference.

> Also I discovered one other minor bug while fixing some stuff in q2pro-rerelease: the Player Setup menu renders a bit funny now.

Nice catch, I didn't notice that. I've simply disabled bloom in menu until this gets fixed (very low priority issue though).

**Comment by Paril** [comment]
Yeah I think you'd only notice the box filtering at low res too. The best place to look is in N64 maps with the ceiling lights; if you view them at steep angles, you can see them shimmer/alias when you move slowly. I'll take a peek when I merge your update in today.

**Comment by Paril** [comment]
I had time this morning to merge your update in - I was initially worried that the box blur would be extra work for barely any benefit, but it reduces the aliasing in the bloom quite a bit. Looks a lot more consistent now, nice work!

**Comment by skullernet** [comment]
Guess I'll merge it too then. Thanks for all suggestions.
